fog/distribution/src/main.rs

```diff
@@ -369,32 +402,68 @@ fn worker_thread_entry(
                 .expect("No fog report url"),
         )
         .expect("Could not parse fog url");
-        // Sometime transactions could not be submitted before tombstone block passed,
-        // so loop until transactions can be submmitted
-        loop {
-            // Got our inputs, construct transaction.
-            let tx = build_tx(
-                &pending_spendable_txouts,
-                to_account,
-                &config,
-                &ledger_db,
-                fog_resolver.clone(),
-                &logger,
-            );
 
-            // Submit tx
-            if submit_tx(txs_created, &conns, &tx, &config, &logger) {
-                txs_created += 1;
-                let mut map = TX_PUB_KEY_TO_ACCOUNT_KEY.lock().unwrap();
-                map.insert(tx.prefix.outputs[0].public_key, to_account.clone());
-                break;
-            } else {
-                //each worker thread should build its own FogResolver,
-                //if submit fails, it should trash and rebuild the FogResolver to ensure it has
-                // fresh fog
-                fog_resolver = build_fog_resolver(&fog_uri, &env, &logger);
-            }
-            log::trace!(logger, "rebuilding failed tx");
+        if let Some(new_fog_resolver) = build_and_submit_transaction(
+            txs_created,
+            pending_spendable_txouts,
+            to_account,
+            &config,
+            &ledger_db,
+            &fog_resolver,
+            &logger,
+            &conns,
+            &env,
+            &fog_uri,
+        ) {
+            fog_resolver = new_fog_resolver;
+        }
+        txs_created += 1;
+    }
+}
+
+/// Builds and submits a transaction to a given FogAccount.
+///
+/// If a transaction submit errors, then we get and use a new FogResolver
+/// to build and submit transactions. In this case, we return this new
+/// FogResolver to the caller so that it can be used in subsequent transactions.
+fn build_and_submit_transaction(
+    txs_created: usize,
+    pending_spendable_txouts: Vec<SpendableTxOut>,
+    to_account: &AccountKey,
+    config: &Config,
+    ledger_db: &LedgerDB,
+    fog_resolver: &FogResolver,
+    logger: &Logger,
+    conns: &[SyncConnection<ThickClient<HardcodedCredentialsProvider>>],
+    env: &Arc<grpcio::Environment>,
+    fog_uri: &FogUri,
+) -> Option<FogResolver> {
+    // Sometimes transactions can not be submitted before the tombstone block
+    // has passed, so loop until transactions can be submmitted
+    loop {
+        let tx = build_tx(
+            &pending_spendable_txouts,
+            to_account,
+            config,
+            ledger_db,
+            fog_resolver.clone(),
```

Comment:
Should this be using the new_fog_resolver.clone()? As is I think the loop just keeps using the old fog resolver. Unless I'm missing something.

Reply:
Yes. I reformatted the method slightly to achieve this.

Reply:
Thanks. Do you think it would make sense to add a test for a fog resolver failing just to make sure that this loop works?

Reply:
I think it's fine to not have a test here- using a new fog resolver is an implementation detail that has the intended behavior of making transfers possible after a failure scenario. IMO we shouldn't be testing implementation details.